src/js/Audit.js

```diff
@@ -218,68 +218,122 @@ axs.Audit.getRulesCannotRun = function(opt_configuration) {
         }
 
         axs.Audit.unsupportedRulesWarningShown = true;
     }
 
+    var auditRules = axs.AuditRules.getActiveRules(configuration);
+
+    // As a performance optimization set the walkDom flag to false if none of the rules needs it.
+    if (!('walkDom' in configuration)) {  // Allow the caller to override this if they KNOW they only need to walk the scope
+        configuration.walkDom = auditRules.some(function(auditRule) {
+            return auditRule.walkDom;
+        });
+    }
+
+    // As a performance optimization set the collectIdRefs flag to false if none of the rules needs it.
+    configuration.collectIdRefs = auditRules.some(function(auditRule) {
+        return auditRule.collectIdRefs;
+    });
+
+    if (!configuration.scope) {
+        configuration.scope = document.documentElement;
+    }
+
+    axs.Audit.collectMatchingElements(configuration, auditRules);
+
+    var results = [];
     for (var i = 0; i < auditRules.length; i++) {
-        var auditRuleName = auditRules[i];
-        var auditRule = axs.AuditRules.getRule(auditRuleName);
-        if (!auditRule)
-            continue; // Shouldn't happen, but fail silently if it does.
-        if (auditRule.disabled)
-            continue;
-        if (!withConsoleApi && auditRule.requiresConsoleAPI)
+        var auditRule = auditRules[i];
+        if (!auditRule.canRun(configuration))
             continue;
+        results.push(auditRule.run(configuration));
+    }
+    return results;
+};
+goog.exportSymbol('axs.Audit.run', axs.Audit.run);
+
+(function() {
+    /**
+     * Runs the element collection phase of the audit.
+     * This performs a complete traversal of the relevant DOM tree.
+     *    Each registered AudtRule is given the opportunity to examine each DOM element
+     *    to determine if it is "relevant" or "interesting".
+     *
+     * Since the DOM may change over time it is recommended to run the Audits in the same event loop as this call.
+     * @param {axs.AuditConfiguration} configuration Configuration for this run.
+     * @param {Array.<axs.AuditRule>} auditRules The active audit rules.
+     */
+    axs.Audit.collectMatchingElements = function(configuration, auditRules) {
+        var rootFlags = {
+            walkDom: configuration.walkDom,
+            collectIdRefs: configuration.collectIdRefs,
+            level: 0,
+            ignoring: {},
+            disabled: false,
+            hidden: false
+        };
+        var root = configuration.walkDom ? document.documentElement : configuration.scope;
+        // Because 'related elements' could occur anywhere in the DOM we need to start at document.documentElement
+        axs.dom.composedTreeSearch(root, null, { preorder: function(element, flags) {
+            if (!flags.inScope)
+                flags.inScope = element === configuration.scope;
```

Comment:
This will need to be undone when leaving the scope - you could add a `postorder` function to achieve this.